### ISSUE TITLE
Complex Number support for torch nn modules

### ISSUE THREAD

**Issue by anjali411** [pytorch/pytorch#46374]
## 🚀 Feature
This issue is meant to roll up the conversation for how PyTorch intends to extend complex number support to nn module.

## Motivation

Complex Neural Networks are quickly becoming an active area of research and it would be useful for the users to be able to create modules with complex valued parameters.

## Pitch

Some of the things intended to be discussed further are:
1. Creating modules with complex valued parameters without having to go via real. for example., `module.to(torch.cfloat)`. The downside of using `module.to(torch.cfloat)` is that the parameters initialized have imaginary value set to 0.
2. Initialization of complex valued parameters. e.g., how should the complex weights be initialized?

@albanD suggested to consider adding Complex versions of the existing modules since weight initialization logic for floating point and complex parameters would be fairly different and keep these modules out-of-tree.

I think we could consider adding a dtype kwarg argument so that user can explicitly specify the kind (complex or floating point) of parameters they would like to have and initialize the parameters accordingly.

**Comment by ezyang** [comment]
nn module initialization has long been a sore point for PyTorch; for example, we still don't have a canonical way to directly initialize modules on CUDA, users have to initialize on CPU and move to CUDA.

**Comment by vadimkantorov** [comment]
Related to @ezyang point: https://github.com/pytorch/pytorch/issues/29523

**Comment by soumickmj** [comment]
Anyone working on it?
I am thinking instead of having direct complex weights, maybe we can work with complex-valued convolutions and other specialized functions? Even if we have direct complex weights in nn modules, maybe would be good to have it in addition?
https://arxiv.org/pdf/1705.09792.pdf This paper talks about the ideas. Also some other papers too. There are some codes available online already (for old way of complex or even their custom way of handling complex) and some not. I'm thinking of polishing them and also finish the not available ones, for the latest PyTorch implementation. I'm also talking about Complex Activation Functions. Will anyone be interested in this? Then we can discuss further on the same.

**Comment by mruberry** [comment]
Following [this](https://github.com/pytorch/pytorch/issues/47052#issuecomment-721241972) comment we are working on supporting complex modules and complex neural networks. However we're not accepting complex activation functions or modules into PyTorch at this time because the field is still new. It would be awesome to see libraries built using PyTorch that did have complex activation functions and modules, however.

**Comment by soumickmj** [comment]
@mruberry Thanks and sorry, I didn't understand one thing. You mentioned you are working on complex modules but said you aren't accepting complex modules. So is that mean PyTorch not accepting complex modules from outside? 
And regarding the work you guys are doing on supporting complex modules and complex neural networks, can I contribute?

**Comment by mruberry** [comment]
We're working on SUPPORTING complex modules. That is, we're working on letting people develop their own complex modules, at the moment. 

You're certainly welcome to contribute to adding more complex support to PyTorch. We have several open complex issues, like [this one for complex loss functions](https://github.com/pytorch/pytorch/issues/46642), with open tasks. Would you be interesting in updating some loss functions to support complex tensors?

**Comment by soumickmj** [comment]
Sure @mruberry. I will look into the loss functions when I get a bit of time. Would love to contribute on that front.
Regarding the complex modules I mentioned, I started developing a package for Complex Modules, following the exact structure of PyTorch. It should be easy for anyone to use these modules as they can use the ditto conventions as PyTorch. 
Its available on my Github pytorch-complex: (https://github.com/soumickmj/pytorch-complex)
If anyone of you interested in Complex Modules, please have a look at the package. I just started working on it today as I need it for my research. Could be buggy at this point. You are most welcome to comment if something is wrong or if something is not working. Also, any other feedbacks will be highly appreciated. Thanks.
cc @ezyang @anjali411 @dylanbespalko @albanD

**Comment by willtop** [comment]
Hello everyone, just want to confirm that now has pytorch been able to support initializing and training complex weights, for say just a simple fully connected neural network (with non-linearities)?

From my understanding, from version 1.7 onwards, the complex numbers were firstly introduced into pytorch, with operations such as matrix multiplications and computing gradients supported, but there didn't seem to be support for building a neural network with complex inputs and weights. Has there been any addition along this direction since 1.7? 

Many thanks for the clarification!

**Comment by soumickmj** [comment]
@willtop Building complex networks are possible as long as the weights are stored and backpropagated as real. You can have a look at my package https://github.com/soumickmj/pytorch-complex  using modules from which you can easily  build complex-valued neural networks. Please feel free to get back to me for any clarifications.

**Comment by willtop** [comment]
@soumickmj Thanks a lot for the information! I will take a look.

**Comment by bonevbs** [comment]
Weighing in on this as this has been a major pain point for me. 

@ezyang has mentioned it before and it would be good to have better ways of controlling `dtype` and `device` of `nn.Modules`. Another pain point is that modules will often have complex and real weights. I have yet to find a good way to call `.double()` and have all my complex tensors cast to complex double precision instead of complex.

The best way at the moment is usually to store everything as real and call `view_as_complex` at the right times. There should be a better way I think.

**Comment by ezyang** [comment]
Hmm, shouldn't we just have a `complex_double` method for that conversion, if that's what you want? Re controlling dtype, maybe we can apply the same playbook as https://github.com/pytorch/pytorch/issues/82296 but it seems kind of delicate, IMO
